### PR TITLE
Interpreter use switch() for main loop

### PR DESCRIPTION
**Pull request by rkimballn1**
The main purpose of this was to add a simple way for a backend to use a switch() statement instead of a bunch of if/else statements.
Some fun stuff is the op.tbl file which is a single list of all ops which is used to generate the two tables needed to make this all happen. The end result is kind of nice with only minimal use of macros.
I confined all of the use of op.tbl into the node_wrapper.?pp files so it is easier to deal with.
I enabled extra error checking on the switch() statement so that all cases MUST be handled.
I also added `const` where I could because, const

### CODE REVIEW

File: src/ngraph/runtime/interpreter/int_backend.hpp [hunk 1]
```diff
@@ -683,69 +734,77 @@ class ngraph::runtime::interpreter::INTBackend : public Backend
...
         }
-        else if (node_op == "Parameter")
-        {
+        case OP_TYPEID::Parameter_TYPEID: { break;
         }
-        else if (node_op == "Pad")
+        case OP_TYPEID::Pad_TYPEID:
         {
-            op::Pad* pad = dynamic_cast<op::Pad*>(&node);
+            const op::Pad* pad = dynamic_cast<const op::Pad*>(&node);
```
Comment (ransford2011) [inline src/ngraph/runtime/interpreter/int_backend.hpp]: `static_cast`?

File: src/ngraph/runtime/interpreter/int_backend.hpp [hunk 2]
```diff
@@ -755,26 +814,29 @@ class ngraph::runtime::interpreter::INTBackend : public Backend
...
                                   args[0]->get_shape(),
                                   out[0]->get_shape(),
                                   product->get_reduction_axes());
+            break;
         }
-        else if (node_op == "Reduce")
+        case OP_TYPEID::Reduce_TYPEID:
         {
-            op::Reduce* reduce = dynamic_cast<op::Reduce*>(&node);
+            const op::Reduce* reduce = dynamic_cast<const op::Reduce*>(&node);
```
Comment (ransford2011) [inline src/ngraph/runtime/interpreter/int_backend.hpp]: `static_cast`?

File: src/ngraph/runtime/interpreter/int_backend.hpp [hunk 3]
```diff
@@ -797,10 +859,11 @@ class ngraph::runtime::interpreter::INTBackend : public Backend
                               node.get_output_shape(0),
                               reduce->get_reduction_axes(),
                               f);
+            break;
         }
-        else if (node_op == "ReduceWindow")
+        case OP_TYPEID::ReduceWindow_TYPEID:
         {
-            op::ReduceWindow* reduce_window = dynamic_cast<op::ReduceWindow*>(&node);
+            const op::ReduceWindow* reduce_window = dynamic_cast<const op::ReduceWindow*>(&node);
```
Comment (ransford2011) [inline src/ngraph/runtime/interpreter/int_backend.hpp]: `static_cast`?

File: src/ngraph/runtime/interpreter/int_backend.hpp [hunk 4]
```diff
@@ -848,35 +914,39 @@ class ngraph::runtime::interpreter::INTBackend : public Backend
                                         slice->get_upper_bounds(),
                                         slice->get_strides(),
                                         out[0]->get_shape());
+            break;
         }
-        else if (node_op == "Reshape")
+        case OP_TYPEID::Reshape_TYPEID:
         {
-            op::Reshape* reshape = dynamic_cast<op::Reshape*>(&node);
+            const op::Reshape* reshape = dynamic_cast<const op::Reshape*>(&node);
```
Comment (ransford2011) [inline src/ngraph/runtime/interpreter/int_backend.hpp]: `static_cast`?

File: src/ngraph/runtime/interpreter/int_backend.hpp [hunk 5]
```diff
@@ -848,35 +914,39 @@ class ngraph::runtime::interpreter::INTBackend : public Backend
...
                                args[0]->get_shape(),
                                reshape->get_input_order(),
                                out[0]->get_shape());
+            break;
         }
-        else if (node_op == "Result")
+        case OP_TYPEID::Result_TYPEID:
         {
-            op::Result* res = dynamic_cast<op::Result*>(&node);
+            const op::Result* res = dynamic_cast<const op::Result*>(&node);
```
Comment (ransford2011) [inline src/ngraph/runtime/interpreter/int_backend.hpp]: `static_cast`?

File: src/ngraph/runtime/interpreter/int_backend.hpp [hunk 6]
```diff
@@ -848,35 +914,39 @@ class ngraph::runtime::interpreter::INTBackend : public Backend
...
             reference::result(args[0]->get_data_ptr<T>(),
                               out[0]->get_data_ptr<T>(),
                               shape_size(res->get_shape()));
+            break;
         }
-        else if (node_op == "Reverse")
+        case OP_TYPEID::Reverse_TYPEID:
         {
-            op::Reverse* reverse = dynamic_cast<op::Reverse*>(&node);
+            const op::Reverse* reverse = dynamic_cast<const op::Reverse*>(&node);
```
Comment (ransford2011) [inline src/ngraph/runtime/interpreter/int_backend.hpp]: `static_cast`?

File: src/ngraph/runtime/interpreter/int_backend.hpp [hunk 7]
```diff
@@ -848,35 +914,39 @@ class ngraph::runtime::interpreter::INTBackend : public Backend
...
                                args[0]->get_shape(),
                                out[0]->get_shape(),
                                reverse->get_reversed_axes());
+            break;
         }
-        else if (node_op == "ReverseSequence")
+        case OP_TYPEID::ReverseSequence_TYPEID:
         {
-            op::ReverseSequence* reverse = dynamic_cast<op::ReverseSequence*>(&node);
+            const op::ReverseSequence* reverse = dynamic_cast<const op::ReverseSequence*>(&node);
```
Comment (ransford2011) [inline src/ngraph/runtime/interpreter/int_backend.hpp]: `static_cast`?

File: src/ngraph/runtime/interpreter/int_backend.hpp [hunk 8]
```diff
@@ -891,19 +961,21 @@ class ngraph::runtime::interpreter::INTBackend : public Backend
...
                                  out[0]->get_element_count());
+            break;
         }
-        else if (node_op == "SelectAndScatter")
+        case OP_TYPEID::SelectAndScatter_TYPEID:
         {
-            ngraph::op::SelectAndScatter* select_and_scatter =
-                dynamic_cast<ngraph::op::SelectAndScatter*>(&node);
+            const ngraph::op::SelectAndScatter* select_and_scatter =
+                dynamic_cast<const ngraph::op::SelectAndScatter*>(&node);
```
Comment (ransford2011) [inline src/ngraph/runtime/interpreter/int_backend.hpp]: `static_cast`?

File: src/ngraph/runtime/interpreter/int_backend.hpp [hunk 9]
```diff
@@ -348,10 +371,11 @@ class ngraph::runtime::interpreter::INTBackend : public Backend
                                             apb->get_padding_below(),
                                             apb->get_padding_above(),
                                             apb->get_include_padding_in_avg_computation());
+            break;
         }
-        else if (node_op == "Broadcast")
+        case OP_TYPEID::Broadcast_TYPEID:
         {
-            op::Broadcast* broadcast = dynamic_cast<op::Broadcast*>(&node);
+            const op::Broadcast* broadcast = dynamic_cast<const op::Broadcast*>(&node);
```
[the remark this answers is not in the document]
Comment (rkimballn1) [reply src/ngraph/runtime/interpreter/int_backend.hpp]: changed all of these

File: src/ngraph/op/op.tbl [hunk 10]
```diff
@@ -0,0 +1,112 @@
+//*****************************************************************************
```
Comment (diyessi) [inline src/ngraph/op/op.tbl]: I think `op_tbl.hpp` would be better for editors that use file extension for syntax selection.

File: src/ngraph/op/op.tbl [hunk 11]
```diff
@@ -0,0 +1,112 @@
...
+// #undef NGRAPH_OP_LIST
+//
+// This sample expands to a list like this:
+// "Abs",
+// "Acos",
+// ...
+//
+// It's that easy. You can use this for fun and profit.
+
+NGRAPH_OP_LIST(Abs)
```
Comment (diyessi) [inline src/ngraph/op/op.tbl]: Should be `NGRAPH_OP` -- these are items of a list, not lists.

File: src/ngraph/runtime/interpreter/int_backend.cpp [hunk 12]
```diff
@@ -164,35 +171,37 @@ bool runtime::interpreter::INTBackend::call(shared_ptr<Function> function,
 
         // get op type
         element::Type type;
-        if (dynamic_pointer_cast<op::util::BinaryElementwiseComparison>(op) ||
-            dynamic_pointer_cast<op::Select>(op))
+        switch (type_id)
         {
+        case OP_TYPEID::Convert_TYPEID:
```
Comment (diyessi) [inline src/ngraph/runtime/interpreter/int_backend.cpp]: `OP_TYPEID::x_TYPEID` seems redundant, and with CamelCase we shouldn't be hitting reserved words, so how about `OP_TYPEID::x`?

File: src/ngraph/op/op.tbl (hunk 10 above)
Comment (rkimballn1) [reply src/ngraph/op/op.tbl]: done

File: src/ngraph/op/op.tbl (hunk 11 above)
Comment (rkimballn1) [reply src/ngraph/op/op.tbl]: done

File: src/ngraph/runtime/interpreter/int_backend.cpp (hunk 12 above)
Comment (rkimballn1) [reply src/ngraph/runtime/interpreter/int_backend.cpp]: done

**Comment by raramer01** [comment]
TESTNOW

**Pull request merged**